internal/unionstore/memdb_art.go

```diff
@@ -165,3 +196,173 @@ func (db *artDBWithContext) SnapshotIterReverse(upper, lower []byte) Iterator {
 func (db *artDBWithContext) SnapshotGetter() Getter {
 	return db.ART.SnapshotGetter()
 }
+
+type snapshotBatchedIter struct {
+	db                    *artDBWithContext
+	snapshotTruncateSeqNo int
+	lower                 []byte
+	upper                 []byte
+	reverse               bool
+	err                   error
+
+	// current batch
+	keys      [][]byte
+	values    [][]byte
+	pos       int
+	batchSize int
+	nextKey   []byte
+
+	// only used to check if the snapshot ever changes between batches. It is not supposed to change.
+	snapshot MemDBCheckpoint
+}
+
+func (db *artDBWithContext) BatchedSnapshotIter(lower, upper []byte, reverse bool) Iterator {
+	if len(db.Stages()) == 0 {
+		logutil.BgLogger().Error("should not use BatchedSnapshotIter for a memdb without any staging buffer")
+	}
+	iter := &snapshotBatchedIter{
+		db:                    db,
+		snapshotTruncateSeqNo: db.SnapshotSeqNo,
+		lower:                 lower,
+		upper:                 upper,
+		reverse:               reverse,
+		batchSize:             32,
+	}
+
+	iter.snapshot = db.GetSnapshot()
+	iter.err = iter.fillBatch()
+	return iter
+}
+
+func (it *snapshotBatchedIter) fillBatch() error {
+	if it.snapshotTruncateSeqNo != it.db.SnapshotSeqNo {
+		return errors.Errorf(
+			"invalid iter: truncation happened, iter's=%d, db's=%d",
+			it.snapshotTruncateSeqNo,
+			it.db.SnapshotSeqNo,
+		)
+	}
+
+	if it.db.GetSnapshot() != it.snapshot {
+		return errors.Errorf(
+			"snapshot changed between batches, expected=%v, actual=%v",
+			it.snapshot,
+			it.db.GetSnapshot(),
+		)
+	}
+
+	it.db.RLock()
+	defer it.db.RUnlock()
```

Comment:
There seems to be possible race condition bettwen the `GetSnapshot` check and `RLock`, it seems we need a CAS like semantics.

Reply:
I reconsidered this and it looks like we don't need to store and compare snapshots at all: `SnapshotSeqNo` is enough to guarantee the "snapshot" doesn't change because of its definition.

Though the race of `snapshotSeqNo` shall be considered another problem